File: mcml-log/src/log.cpp
#include "log.hpp"

void Log::Init(const std::string &local)
{
    log_level_ = LogLevel::LOG_INFO | LogLevel::LOG_WARN | LogLevel::LOG_ERROR | LogLevel::LOG_FATAL;

    log_file_ = std::filesystem::path(local) / "log.logs";
    log_stream_.open(log_file_, std::ios::app);
    
    if (log_stream_.is_open())
    {
        init_ = true;
        run_ = true;
        log_thread_ = new CoreThread("mcml log", Run);
    }
    else
    {
        init_ = false;
    }
}

void Log::Stop()
{
    if (run_)
    {
        log_semaphore_.release();
    }
    run_ = false;
}

void Log::PutLog(LogLevel level, std::string* log)
{
    if (log_level_ | level)
    {
        LogItem* item = new LogItem();
        item->level = level;
        item->log = log;

        log_queue_.push(item);
        log_semaphore_.release();
    }
}

void Log::Info(const std::string &str)
{
    PutLog(LogLevel::LOG_INFO, new std::string(str));
}

void Log::Warn(const std::string &str)
{
    PutLog(LogLevel::LOG_WARN, new std::string(str));
}

void Log::Error(const std::string &str) 
{
    PutLog(LogLevel::LOG_ERROR, new std::string(str));
}

void Log::Fatal(const std::string &str) 
{
    PutLog(LogLevel::LOG_FATAL, new std::string(str));
}

void Log::Run()
{
    while (run_)
    {
        log_semaphore_.acquire();
        LogItem* item;
        while (log_queue_.pop(item))
        {
            time_t nowtime;
            time(&nowtime);
            tm* p = localtime(&nowtime);
            std::string time = std::format("{}/{}/{} {}-{}-{}", p->tm_year, p->tm_mon + 1, p->tm_mday,
                p->tm_hour, p->tm_min, p->tm_sec);
            log_stream_ << '[' << GetLogLevelName(item->level) << ']' << '[' << time << ']' << item->log->c_str() << "\r\n";
        }
    }
}

File: mcml-log/src/log.rs
use crossbeam_queue::SegQueue;
use semrs::Semaphore;

use std::{
    fs::{File, OpenOptions},
    io::{BufWriter, Write},
    path::PathBuf,
    sync::{
        Arc, Mutex, OnceLock, RwLock,
        atomic::{AtomicBool, Ordering},
    },
    thread::{self},
};

use crate::log_item::{LogItem, LogLevel};

static LOG_QUEUE: RwLock<SegQueue<LogItem>> = RwLock::new(SegQueue::new());
static LOG_LOCAL: OnceLock<PathBuf> = OnceLock::new();
static LOG_FILE: OnceLock<Mutex<BufWriter<File>>> = OnceLock::new();
static LOG_RUN: AtomicBool = AtomicBool::new(true);
static LOG_SEM: OnceLock<Arc<Semaphore>> = OnceLock::new();

pub fn start(local: String) {
    LOG_SEM.get_or_init(|| Arc::new(Semaphore::new(0)));

    let log_path = PathBuf::from(local).join("logs.log");

    let file = match OpenOptions::new()
        .create(true)
        .append(true)
        .write(true)
        .open(&log_path)
    {
        Ok(f) => f,
        Err(e) => {
            panic!("log open fail: {} - {}", log_path.display(), e);
        }
    };

    if LOG_LOCAL.set(log_path).is_err() {
        panic!("LOG_LOCAL fail");
    }

    if LOG_FILE.set(Mutex::new(BufWriter::new(file))).is_err() {
        panic!("LOG_FILE fail");
    }

    let thread = thread::Builder::new()
        .name("Log thread".into())
        .spawn(|| run());

    if thread.is_err() {
        panic!("Log Thread start fail")
    }
}

pub fn stop() {
    LOG_RUN.store(false, Ordering::Release);
}

fn save() {
    let log = LOG_QUEUE.read().unwrap();
    let mut file = LOG_FILE.get().unwrap().lock().unwrap();

    while !log.is_empty() {
        let item = log.pop();
        if item.is_some() {
            let item1 = item.unwrap();
            file.write_fmt(format_args!(
                "[{}][{}]{}{}",
                item1.get_time(),
                item1.get_level(),
                item1.log,
                if cfg!(windows) { "\r\n" } else { "\n" }
            ))
            .unwrap();
        }
    }
}

pub fn run() {
    while LOG_RUN.load(Ordering::Acquire) {
        LOG_SEM.get().unwrap().down();

        save();
    }

    save();
}

pub fn info(text: String) {
    LOG_QUEUE
        .write()
        .unwrap()
        .push(LogItem::new(text, LogLevel::Info));
    LOG_SEM.get().unwrap().up();
}

pub fn warn(text: String) {
    LOG_QUEUE
        .write()
        .unwrap()
        .push(LogItem::new(text, LogLevel::Warn));
    LOG_SEM.get().unwrap().up();
}

pub fn error(text: String) {
    LOG_QUEUE
        .write()
        .unwrap()
        .push(LogItem::new(text, LogLevel::Error));
    LOG_SEM.get().unwrap().up();
}

pub fn failt(text: String) {
    LOG_QUEUE
        .write()
        .unwrap()
        .push(LogItem::new(text, LogLevel::Fault));
    LOG_SEM.get().unwrap().up();
}
